lib/compat/wordpress-6.0/class-wp-theme-json-gutenberg.php

```diff
@@ -161,4 +161,154 @@ public function get_data() {
+		 * We rather replace the existing with the incoming value, if it exists.
+		 * This happens, for example, when we merge data from theme.json upon existing
+		 * theme supports or when we merge anything coming from the same source twice.
+		 * This is the case of color.palette, color.gradients, color.duotone,
+		 * typography.fontSizes, or typography.fontFamilies.
+		 *
+		 * Additionally, for some preset types, we also want to make sure the
+		 * values they introduce don't conflict with default values. We do so
+		 * by checking the incoming slugs for theme presets and compare them
+		 * with the equivalent default presets: if a slug is present as a default
+		 * we remove it from the theme presets.
+		 */
+		$nodes        = static::get_setting_nodes( $replacements );
+		$slugs_global = static::get_default_slugs( $base, array( 'settings' ) );
+		foreach ( $nodes as $node ) {
+			$slugs_node = static::get_default_slugs( $base, $node['path'] );
+			$slugs      = array_merge_recursive( $slugs_global, $slugs_node );
+
+			// Replace the spacing.units.
+			$path    = array_merge( $node['path'], array( 'spacing', 'units' ) );
+			$content = _wp_array_get( $replacements, $path, null );
+			if ( isset( $content ) ) {
+				_wp_array_set( $merged_data, $path, $content );
+			}
+
+			// Replace the presets.
+			foreach ( static::PRESETS_METADATA as $preset ) {
+				if ( array_key_exists( 'settings', $base ) ) {
+					$override_preset = ! static::get_metadata_boolean( $base['settings'], $preset['prevent_override'], true );
+				}
+
+				foreach ( static::VALID_ORIGINS as $origin ) {
+					$base_path = array_merge( $node['path'], $preset['path'] );
+					$path      = array_merge( $base_path, array( $origin ) );
+					$content   = _wp_array_get( $replacements, $path, null );
+					if ( ! isset( $content ) ) {
+						continue;
+					}
+
+					if ( 'theme' === $origin && $preset['use_default_names'] ) {
+						foreach ( $content as &$item ) {
+							if ( ! array_key_exists( 'name', $item ) ) {
+								$name = static::get_name_from_array( $base, $item['slug'], $base_path );
+								if ( null !== $name ) {
+									$item['name'] = $name;
+								}
+							}
+						}
+					}
+
+					if (
+						( 'theme' !== $origin ) ||
+						( 'theme' === $origin && $override_preset )
+					) {
+						_wp_array_set( $base, $path, $content );
+					} else {
+						$slugs_for_preset = _wp_array_get( $slugs, $preset['path'], array() );
+						$content          = static::filter_slugs( $content, $slugs_for_preset );
+						_wp_array_set( $base, $path, $content );
+					}
+				}
+			}
+		}
+
+		return $merged_data;
+	}
+
+	/**
+	 * Gets the result of merging the theme's theme.json and the user settings
+	 *
+	 * @since 6.0.0
+	 *
+	 * @return array The result of the merge
+	 */
+	public static function get_merged_theme_json() {
+		// Get the user data
+		$user_data = WP_Theme_JSON_Resolver_Gutenberg::get_user_data();
```

Comment:
We've been trying to make the `WP_Theme_JSON` class agnostic from where the data comes from and from WordPress APIs (`get_stylesheet`, etc), while the `WP_Theme_JSON_Resolver` responsibilitiy is to mangle data together.